File: src/hackerrank/functional_programming/recursion/StringCompression.scala
package hackerrank.functional_programming.recursion

import java.util.Scanner

/**
  * Created by Aleksandr on 17/02/2017.
  * Project Solutions
  *
  * https://www.hackerrank.com/challenges/string-compression
  */
object StringCompression {
  def main(args: Array[String]) {
    val in = new Scanner(System.in)
    val s = in.next()

    val sb = new StringBuilder()
    sb.append(s.charAt(0))
    goDeeper(sb, s.toList.drop(1), 1)

    println(sb.toString())
  }

  def goDeeper(sb: StringBuilder, list: List[Char], count: Int): Unit = {
    if(list.isEmpty) {
      if(count > 1) {
        sb.append(count)
      }
      return
    }

    if(sb.last == list.head) {
      goDeeper(sb, list.drop(1), count + 1)
    } else {
      if(count > 1) {
        sb.append(count)
      }
      goDeeper(sb.append(list.head), list.drop(1), 1)
    }
  }
}

File: src/leetcode/StringCompression.java
package leetcode;

/**
 * Created by Aleksandr on 14/12/2017.
 * Project Solutions
 * <p>
 * https://leetcode.com/problems/string-compression
 */
public class StringCompression {
    public int compress(char[] chars) {
        if(chars == null || chars.length == 0) {
            return 0;
        }

        int start = 0, end = 0, compressed = 0;
        while(end < chars.length) {
            if(chars[start] == chars[end]) {
                end++;
            } else {
                compressed += compress(chars, compressed, start, end);
                start = end;
            }
        }

        compressed += compress(chars, compressed, start, end);

        return compressed;
    }

    private int compress(char[] chars, int writeIndex, int start, int end) {
        final String s = String.valueOf(end - start);
        chars[writeIndex] = chars[start];
        if(end - start >= 2) {
            for(int i = 0; i < s.length(); i++) {
                chars[writeIndex + i + 1] = s.charAt(i);
            }
            return 1 + s.length();
        } else {
            return 1;
        }
    }
}
